mp4parse_capi/src/lib.rs

```diff
@@ -695,84 +883,123 @@ fn create_sample_table(track: &Track) -> Option<Vec<mp4parse_indice>> {
-                        // TODO: 'elst' is optional box, we need to check 'elst' for start composition time
-                        //       if it exists.
-                        sample.start_decode = decode_time;
-                        sample.start_composition = decode_time;
-                        sample.end_composition = end_composition;
-                        // TODO: 'ctts' is optional box, we need to check ctts for composition time
-                        //       if it exists.
-                    },
-                    _ => {
-                        return None;
-                    }
-                }
+    let ctts_iter = match &track.ctts {
+        &Some(ref v) => Some(v.samples.as_slice().iter()),
+        _ => None,
+    };
+
+    let mut ctts_offset_iter = TimeOffsetIterator {
+        cur_sample_range: (0 .. 0),
+        cur_offset: 0,
+        ctts_iter: ctts_iter,
+    };
+
+    let mut stts_iter = TimeToSampleIteraor {
+        cur_sample_count: (0 .. 0),
+        cur_sample_delta: 0,
+        stts_iter: stts.samples.as_slice().iter(),
+    };
+
+    // sum_delta is the sum of stts_iter delta.
+    // According to sepc:
+    //      decode time => DT(n) = DT(n-1) + STTS(n)
+    //      composition time => CT(n) = DT(n) + CTTS(n)
+    // Note:
+    //      composition time needs to add the track offset time from 'elst' table.
+    let mut sum_delta = PresentationTime::new(0, timescale);
+    for sample in sample_table.as_mut_slice() {
+        let decode_time = sum_delta.to_us();
+        sum_delta.time += stts_iter.next_delta() as i64;
+
+        // ctts_offset is the current sample offset time.
+        let ctts_offset = PresentationTime::new(ctts_offset_iter.next_offset_time(), timescale);
+
+        let start_composition = (decode_time + ctts_offset.to_us() + track_offset_time) as u64;
+        let end_composition = (sum_delta.to_us() + ctts_offset.to_us() + track_offset_time) as u64;
+
+        sample.start_decode = decode_time as u64;
+        sample.start_composition = start_composition;
+        sample.end_composition = end_composition;
+    }
+
+    // Correct composition end time due to 'ctts' causes composition time re-ordering.
+    //
+    // Composition end time is not in specification. However, gecko needs it, so we need to
+    // calculate to correct the composition end time.
+    if track.ctts.is_some() {
+        // Create an index table refers to sample_table and sorted by start_composisiton time.
+        let mut sort_table = Vec::new();
+        sort_table.reserve(sample_table.len());
+        for i in 0 .. sample_table.len() {
+            sort_table.push(i);
+        }
+
+        sort_table.sort_by_key(|i| {
+            match sample_table.get(*i) {
+                Some(v) => {
+                    v.start_composition
+                },
+                _ => 0,
             }
+        });
+
+        let iter = sort_table.iter().peekable();
```

Comment:
Do you need to call `.peekable()` here? You never call `iter.peek()` and the length should be the same for both Iterators.